src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XpathIteratorUtil.java

```diff
@@ -0,0 +1,141 @@
+// Copyright (C) 2001-2021 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.internal.utils;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.puppycrawl.tools.checkstyle.DetailAstImpl;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.xpath.RootNode;
+import net.sf.saxon.om.NodeInfo;
+
+/**
+ * XpathIteratorUtil.
+ *
+ * @noinspection ClassOnlyUsedInOnePackage
+ */
+public final class XpathIteratorUtil {
+
+    /**
+     * Map from {@link NodeInfo#getLocalPart()} to {@link NodeInfo}.
+     */
+    private static final Map<String, NodeInfo> NODES_BY_LOCAL_PART = new HashMap<>();
+
+    static {
+        fillMap(new RootNode(createTree()));
+    }
+
+    private XpathIteratorUtil() {
+    }
+
+    /**
+     * Finds node by its {@link NodeInfo#getLocalPart()} value.
+     *
+     * @param localPart text to be searched
+     *
+     * @return corresponding node
+     */
+    public static NodeInfo findNode(String localPart) {
+        return NODES_BY_LOCAL_PART.get(localPart);
+    }
+
+    /**
+     * Iterates Xpath tree nodes and fills the map.
+     *
+     * @param cur current {@link NodeInfo} node
+     */
+    private static void fillMap(NodeInfo cur) {
+        NODES_BY_LOCAL_PART.put(cur.getLocalPart(), cur);
+        for (NodeInfo node : cur.children()) {
+            fillMap(node);
+        }
+    }
+
+    /**
+     * Creates AST tree for iteration tests and returns root.
+     * <p>
+     * PACKAGE_DEF -> package [1:0]
+     * |--ANNOTATIONS -> ANNOTATIONS [1:8]
+     *     `--ANNOTATION_DEF -> ANNOTATIONS [1:8]
+     * `--CLASS_DEF -> CLASS_DEF [3:0]
+     *     `--OBJBLOCK -> OBJBLOCK [3:19]
+     *         |--LCURLY -> { [3:19]
+     *         |--METHOD_DEF -> METHOD_DEF [4:4]
+     *         |   |--PARAMETERS -> PARAMETERS [4:31]
+     *         |   `--SLIST -> { [4:33]
+     *         `--RCURLY -> } [8:0]
+     * </p>
+     *
+     * @return {@link DetailAST} object.
+     */
+    private static DetailAST createTree() {
```

Comment:
Could we parse a file and get resulting AST here? It would be good to restrict usage of `DetailAstImpl` as much as we can. Also, AST in comment is not consistent with nodes created in method ( `PARAMETERS` and `SLIST` nodes)

Reply:
> Could we parse a file and get resulting AST here? 

Ideally yes, but it is harder to implement. 

I used simplified ast tree to test iteration order. See, I have map `NODES_BY_LOCAL_PART`, where key is name of node, I assume it will be unique for my small tree, which makes it impossible to use with real files, since `IDENT`, `MODIFIERS`, `SLIST`, `RCURLY` can occur several times. 

I haven't found good solution, except xpath-querying node by name, line and column number everytime. It seems overcomplicated :( 

If you have suggestion, please write

Reply:
> Also, AST in comment is not consistent with nodes created in method ( PARAMETERS and SLIST nodes)

done